File: src/main/java/com/maybeitssquid/rotatingsecrets/CredentialRotationException.java
package com.maybeitssquid.rotatingsecrets;

/**
 * Exception thrown when credential rotation fails.
 *
 * <p>This exception wraps underlying errors that occur during the credential
 * rotation process, such as failures to update connection pool credentials
 * or refresh pool connections.</p>
 *
 * @see CredentialsProviderService
 * @see UpdatableCredential
 */
public class CredentialRotationException extends RuntimeException {

    /**
     * Creates a new credential rotation exception with the specified message and cause.
     *
     * @param message detailed description of the rotation failure
     * @param cause   the underlying exception that caused the failure
     */
    public CredentialRotationException(String message, Throwable cause) {
        super(message, cause);
    }
}

File: src/dotnet/MaybeItsSquid.RotatingSecrets/Core/CredentialRotationException.cs
namespace MaybeItsSquid.RotatingSecrets.Core;

/// <summary>
/// Exception thrown when credential rotation operations fail.
/// </summary>
public class CredentialRotationException : Exception
{
    /// <summary>
    /// Initializes a new instance of the <see cref="CredentialRotationException"/> class.
    /// </summary>
    public CredentialRotationException()
    {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="CredentialRotationException"/> class
    /// with a specified error message.
    /// </summary>
    /// <param name="message">The message that describes the error.</param>
    public CredentialRotationException(string message)
        : base(message)
    {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="CredentialRotationException"/> class
    /// with a specified error message and inner exception.
    /// </summary>
    /// <param name="message">The message that describes the error.</param>
    /// <param name="innerException">The exception that caused this exception.</param>
    public CredentialRotationException(string message, Exception innerException)
        : base(message, innerException)
    {
    }
}
